turbopack/crates/turbo-tasks-backend/src/kv_backing_storage.rs

```diff
@@ -648,3 +730,80 @@ fn serialize(task: TaskId, data: &Vec<CachedDataItem>) -> Result<SmallVec<[u8; 1
         }
     })
 }
+
+/// Store the strings in the database and return the global ids.
+fn save_strings_serial<'a>(
+    batch: &mut impl SerialWriteBatch<'a>,
+    strings: &RcStrToLocalId,
+) -> Result<LocalIdToGlobalId> {
+    let mut global_ids = Vec::new();
+    for s in strings.0.iter() {
+        let (global_id, is_new) = get_string_id(batch, s)?;
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+/// Store the strings in the database and return the global ids.
+fn save_strings_concurrent<'a>(
+    batch: &impl ConcurrentWriteBatch<'a>,
+    strings: &RcStrToLocalId,
+) -> Result<LocalIdToGlobalId> {
+    let mut global_ids = Vec::new();
+    for s in strings.0.iter() {
+        let (global_id, is_new) = get_string_id(batch, s)?;
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+
+static STRING_INTERN_ID: OnceLock<AtomicU32> = OnceLock::new();
+
+/// Returns `(global_id, is_new)`
+fn get_string_id<'a>(batch: &impl BaseWriteBatch<'a>, s: &RcStr) -> Result<(u32, bool)> {
+    let original = batch.get(KeySpace::StringInternMap, s.as_bytes())?;
+
+    if let Some(bytes) = original {
+        let global_id = as_u32(bytes)?;
+        return Ok((global_id, false));
+    }
+
+    let Some(bytes) = batch.get(KeySpace::Infra, IntKey::new(META_KEY_STRING_ID).as_ref())? else {
+        return Ok((0, true));
+    };
```

Comment:
Only restore that once during initialization of the backing storage and only operate on the atomic here.

In every snap_snapshot, store it once to the DB.

We have a similar thing with `next_task_id`